float32 latent_x
float32 latent_y